pkg/component/operator/audio/v0/task_detect_voice_activity.go

```diff
@@ -0,0 +1,130 @@
+package audio
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/go-audio/audio"
+	"github.com/go-audio/wav"
+	"github.com/mjibson/go-dsp/fft"
+	"github.com/streamer45/silero-vad-go/speech"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	"github.com/instill-ai/pipeline-backend/pkg/component/base"
+)
+
+type detectVoiceActivityInput struct {
+	Audio                Audio `json:"audio"`
+	MinSilenceDurationMs int   `json:"min-silence-duration-ms"`
+	SpeechPadMs          int   `json:"speech-pad-ms"`
+}
+
+type detectVoiceActivityOutput struct {
+	Segments []segment `json:"segments"`
+}
+
+func detectVoiceActivity(input *structpb.Struct, job *base.Job, ctx context.Context) (*structpb.Struct, error) {
+	inputStruct := &detectVoiceActivityInput{}
+	if err := base.ConvertFromStructpb(input, &inputStruct); err != nil {
+		return nil, fmt.Errorf("converting input to struct: %v", err)
+	}
+
+	buf, err := base64.StdEncoding.DecodeString(base.TrimBase64Mime(string(inputStruct.Audio)))
+	if err != nil {
+		return nil, err
+	}
+
+	dec := wav.NewDecoder(bytes.NewReader(buf))
+	if !dec.IsValidFile() {
+		return nil, fmt.Errorf("invalid WAV file")
+	}
+
+	// Create an IntBuffer
+	audioBuf := &audio.IntBuffer{
+		Format: &audio.Format{
+			NumChannels: int(dec.NumChans),
+			SampleRate:  int(dec.SampleRate),
+		},
+		Data:           make([]int, len(buf)),
+		SourceBitDepth: int(dec.BitDepth),
+	}
+
+	// Read the full audio buffer
+	_, err = dec.PCMBuffer(audioBuf)
+	if err != nil {
+		return nil, fmt.Errorf("error reading audio data: %w", err)
+	}
+
+	if audioBuf.Format.NumChannels > numChannel {
+		audioBuf = toMono(audioBuf)
+	}
+
+	resampledData := resample(audioBuf.AsFloatBuffer().Data, dec.SampleRate, sampleRate)
+	audioBuf.Format.SampleRate = sampleRate
+	audioBuf.Data = resampledData
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		return nil, fmt.Errorf("get current working directory: %w", err)
+	}
+
+	sd, err := speech.NewDetector(speech.DetectorConfig{
+		ModelPath:            filepath.Join(cwd, "model/silero_vad.onnx"),
+		SampleRate:           16000,
+		Threshold:            0.5,
+		MinSilenceDurationMs: inputStruct.MinSilenceDurationMs,
+		SpeechPadMs:          inputStruct.SpeechPadMs,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("detect voice activity: %w", err)
+	}
+
+	segments, err := sd.Detect(audioBuf.AsFloat32Buffer().Data)
+	if err != nil {
+		return nil, fmt.Errorf("detect voice activity: %w", err)
+	}
+
+	if err := sd.Destroy(); err != nil {
```

Comment:
❓ Can we use `defer` for this?

Reply:
We actually want to destroy the model instance right after the inference has been performed.